### ISSUE TITLE
[Design] Uptime Real Browser Test Run Test Details

### ISSUE THREAD

**Issue by drewpost** [elastic/uptime#250]
**User stories**
As a user of real browser based monitors
I want to be able to view a page load waterfall and filmstrip of the page load
So that I can understand what to load experience was of every object on the page
and how it impacted the perceived performance
and so I can troubleshoot errors

**List known (technical) restrictions and requirements**
Ideal Implementation

A view per step of the test where there is a full page load
Available from the monitor detail page from each test run
A link to most recent test run from overview page
Filmstrip view linked to waterfall scrubber
Shows different MIME types
overlays of key page load events
Objects with errors (ie returning 403, 404) highlighted in red
Shows CPU utilisation along load
Shows bandwidth along load
Shows browser main thread along load
has a connection view for each domain
Shows details for each individual request
Shows headers and responses for each request
It should be REALLY nice to look at and use (people will spend a lot of time in this view. We should try to make it delightful to look at and use!)

If in doubt, don’t hesitate to reach out to the `#observability-design` Slack channel.

**Comment by drewpost** [comment]
https://github.com/elastic/uptime-dev/issues/22

**Comment by katrin-freihofner** [comment]
[Figma file](https://www.figma.com/file/VNNDSawIv5crAuOBk3xCzJ/Synthetic-monitoring?node-id=1%3A103)

**Comment by katrin-freihofner** [comment]
Engineering ticket: https://github.com/elastic/kibana/issues/80162

**Comment by katrin-freihofner** [comment]
@drewpost could you please add a list of:
- MIME types
- errors/response codes
- key page load events
we should visualize?

**Comment by drewpost** [comment]
**MIME Types:**
application
audio
example
font
image
model
text
video
plain text
css
html
javascript
from-data

Errors/response codes should handle all HTTP response codes

Key load events:
start render
First Paint
DOM Interactive
DOM content loaded
onLoad
Document Complete

We should also allow for users to choose additional markers (ie if a company is using custom marks/measures. for things like hero images, we should support showing these.

cc @paulb-elastic @vigneshshanmugam anything you'd think I'm missing above or that you would do differently?

**Comment by katrin-freihofner** [comment]
> **MIME Types:**
> application
> audio
> example
> font
> image
> model
> text
> video
> plain text
> css
> html
> javascript
> from-data

I don't think it makes sense to have different colors for each of them. Should we even pick colors for mime types? Can we group them?

**Comment by katrin-freihofner** [comment]
> Errors/response codes should handle all HTTP response codes

Yes, but in the waterfall we are only indicating 4xx and 5xx, right?

**Comment by katrin-freihofner** [comment]
> We should also allow for users to choose additional markers (ie if a company is using custom marks/measures. for things like hero images, we should support showing these.

That is a new feature and not something I see in this initial version. Let's put it in a separate ticket.

**Comment by drewpost** [comment]
> > Errors/response codes should handle all HTTP response codes
> 
> Yes, but in the waterfall we are only indicating 4xx and 5xx, right?

All 4xx and 5xx. I'd also include 301, 307 and 308 as they all indicate the URL for the object you have isn't correct and it's introducing a performance penalty when that incorrect URL is being called and has to redirect.

**Comment by drewpost** [comment]
> > We should also allow for users to choose additional markers (ie if a company is using custom marks/measures. for things like hero images, we should support showing these.
> 
> That is a new feature and not something I see in this initial version. Let's put it in a separate ticket.

@vigneshshanmugam Does the agent capture custom marks and measures today? If so, are they stored in a specific field? I'm wondering if, for v1, we could have it automatically display vertical markers for any values stored in those fields?

**Comment by drewpost** [comment]
> > **MIME Types:**
> > application
> > audio
> > example
> > font
> > image
> > model
> > text
> > video
> > plain text
> > css
> > html
> > javascript
> > from-data
> 
> I don't think it makes sense to have different colors for each of them. Should we even pick colors for mime types? Can we group them?

I think we should do fixed colours for MIME types. I hear you on there being too many. 

HTML
Javascript
CSS
Font
Media (audio and video types grouped)
Other (groups everything else?)

@paulb-elastic Can I get your input on this one?

**Comment by paulb-elastic** [comment]
Here are my thoughts.

### MIME Types

I'd agree to try to simplify the amount of information, colour and legends, we should show fewer MIME types, such as:
* HTML
* JS
* CSS
* Font
* Images
* Everything else


### HTTP Status Codes

I think it's useful to show them all (including 200s, as it's often useful to spot a 200 vs a 201 for example).  We can also then callout those that are **error codes** (maybe colouring the URL/code red, or some other visual cue).

I would think of error codes being:
* 4xx
  * 401 might be something we change in the future, as you'll sometimes get a 401 response which then triggers a subsequent authorisation request, so doesn't always mean something has gone wrong
* 5xx

I get your point about redirects @drewpost (they slow things down), but I don't typically think of them as errors.  Maybe there should be another colour (or visual cue) for *less performant* requests.  Initially, we could use that for things like redirects, but we could also use the same approach in the future with other optimisation suggestions (e.g. non compressed text resources).


### Timing points

I think it's useful to have some of the more traditional (legacy) timing points shown as these are still used by some, when looking at performance, although I think we should also look to add Core Web Vitals to Synthetics as well (or their lab equivalent, where we can't).  These may not all be available yet, but we could consider adding these in the future if not:

* First Contentful Paint (FCP)
* Core Web Vital: Largest Contentful Paint (LCP)
* Core Web Vital: Cumulative Layout Shift (CLS)
* (Sort of) Core Web Vital: Total Blocking Time (TBT) in place of FID as this is a synthetic test
* Maybe - open to thoughts on these as they are typically more legacy timing points now, but may be more familiar for some users
  * onLoad (document complete)
  * DOMContentLoaded
* Custom timing points
  * May not have these yet, but a useful addition in the future, where users instrument their site with specific timing points they want to track

**Comment by vigneshshanmugam** [comment]
> Does the agent capture custom marks and measures today? If so, are they stored in a specific field? I'm wondering if, for v1, we could have it automatically display vertical markers for any values stored in those fields?

Not right now, we are not capturing any performance metrics data(Nav Timing, Paint Metrics, User Timing, Long tasks etc) right now in the synthetics agent. Seeing this issue, It's time to prioritise and start capturing from the agent if we are going to showcase them in the UI. 

@drewpost For the metrics part, IMO we should align with the existing RUM stuff to make the UX page in consistent with the Synthetics run and expand on the other metrics going forward. That means we start surfacing
+ FCP
+ LCP
+ CLS
+ TBT - We capture FID and TBT in RUM agent, however TBT is only till the page load event not throughout the session which is less useful
+ User timing marks/measures

We can expand on the other metrics like Time to Interactive(TTI), Speed Index later on. Thoughts?

**Comment by drewpost** [comment]
OK - 100% agree with Paul and Vignesh. 

MIME Types

HTML
JS
CSS
Font
Images
Everything else

HTTP Status Codes

All result codes that are not one of the below are "normal" and are displayed with the object URL 
PLUS
4xx - errors
401 might be something we change in the future, as you'll sometimes get a 401 response which then triggers a subsequent authorisation request, so doesn't always mean something has gone wrong
5xx - errors
Low performant requests (redirects in v1)

Timing points

First Contentful Paint (FCP)
Core Web Vital: Largest Contentful Paint (LCP)
Core Web Vital: Cumulative Layout Shift (CLS)
Core Web Vital: Total Blocking Time (TBT) in place of FID as this is a synthetic test
onLoad (document complete)
DOMContentLoaded
Custom timing points

cc @katrin-freihofner
